File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Libya.m
//
//  UIImage+Libya.m
//  CountriesFlags
//

#import "Countries/UIImage+Libya.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (Libya)

+ (UIImage * _Nullable)flagLibyaWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:35.0f/255.0f
                                          green:158.0f/255.0f
                                           blue:70.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:231.0f/255.0f
                                        green:0.0f
                                         blue:19.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorBlack = [UIColor blackColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlack setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height / 4.0f));
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 4.0f * 3.0f, size.width, size.height / 4.0f));

    [colorWhite setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGSizeMidPoint(size)
                                    radius:size.height / 8.0f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorBlack setFill];
    [[UIBezierPath bezierPathWithArcCenter:(CGPointMake(size.width * 0.512f, size.height / 2.0f))
                                    radius:size.height / 9.15f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];

    [colorWhite setFill];
    CGPoint starCenter = CGPointMake(size.width * 0.577f, size.height / 2.0f);
    UIBezierPath* path = [UIBezierPath starAtPoint:starCenter
                                            radius:size.height / 11.0
                                              type:FlagStarTypeFivePointed];
    [path rotateToAngle:-18.0f center:starCenter];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Libya.swift
//
//  UIImage+Libya.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://simple.wikipedia.org/wiki/Flag_of_Libya "wikipedia"
     
     Generates the flag of Libya.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Libya.
     */
    internal static func flagLibya(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 35.0 / 255.0,
                                 green: 158.0 / 255.0,
                                 blue: 70.0 / 255.0,
                                 alpha: 1.0)
        let colorRed = UIColor(red: 231.0 / 255.0,
                               green: 0.0,
                               blue: 19.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white
        let colorBlack = UIColor.black

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorBlack.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero,
                          size: CGSize(width: size.width, height: size.height / 4.0)))

        colorGreen.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 4.0 * 3.0),
                          size: CGSize(width: size.width, height: size.height / 4.0)))

        colorWhite.setFill()
        UIBezierPath(arcCenter: size.midPoint,
                     radius: size.height / 8.0,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorBlack.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.width * 0.512, y: size.height / 2.0),
                     radius: size.height / 9.15,
                     startAngle: 0.0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorWhite.setFill()
        let starCenter = CGPoint(x: size.width * 0.577, y: size.height / 2.0)
        let path = UIBezierPath.star(at: starCenter, radius: size.height / 11.0)
        path.rotate(to: -18.0, around: starCenter)
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
